### DISCUSSION TITLE
Getting 5000ms Timeout error while connecting Postgres

### DISCUSSION THREAD

**Discussion by manthan-sf** [loopbackio/loopback-next#7182] [Q&A]
`Error: Timeout in connecting after 5000 ms
    at Timeout._onTimeout (/home/manthan/Desktop/temp/temp/node_modules/loopback-datasource-juggler/lib/datasource.js:2654:10)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)`

But I also made a js file and tried to make  connection there to check any issue with  postgres , 
![Screenshot from 2020-05-26 11-59-15](https://user-images.githubusercontent.com/53930889/82868900-0694cc80-9f4b-11ea-9af8-ab51e5985ab6.png)

Package.json -> 
`
{
  "name": "temp",
  "version": "1.0.0",
  "description": "temp",
  "keywords": [
    "loopback-application",
    "loopback"
  ],
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "build": "npm run clean && lb-tsc",
    "build:watch": "lb-tsc --watch",
    "lint": "npm run prettier:check && npm run eslint",
    "lint:fix": "npm run eslint:fix && npm run prettier:fix",
    "prettier:cli": "lb-prettier \"**/*.ts\" \"**/*.js\"",
    "prettier:check": "npm run prettier:cli -- -l",
    "prettier:fix": "npm run prettier:cli -- --write",
    "eslint": "lb-eslint --report-unused-disable-directives .",
    "eslint:fix": "npm run eslint -- --fix",
    "pretest": "npm run clean && npm run build",
    "test": "lb-mocha --allow-console-logs \"dist/__tests__\"",
    "posttest": "npm run lint",
    "test:dev": "lb-mocha --allow-console-logs dist/__tests__/**/*.js && npm run posttest",
    "docker:build": "docker build -t temp .",
    "docker:run": "docker run -p 3000:3000 -d temp",
    "migrate": "node ./dist/migrate",
    "prestart": "npm run build",
    "start": "node -r source-map-support/register .",
    "clean": "lb-clean dist *.tsbuildinfo .eslintcache"
  },
  "repository": {
    "type": "git"
  },
  "author": "",
  "license": "",
  "files": [
    "README.md",
    "dist",
    "src",
    "!*/__tests__"
  ],
  "dependencies": {
    "@loopback/boot": "^2.3.1",
    "@loopback/context": "^3.8.1",
    "@loopback/core": "^2.7.0",
    "@loopback/openapi-v3": "^3.4.1",
    "@loopback/repository": "^2.5.1",
    "@loopback/rest": "^5.0.1",
    "@loopback/rest-explorer": "^2.2.2",
    "@loopback/service-proxy": "^2.3.0",
    "@types/pg": "^7.14.3",
    "loopback-connector-postgresql": "^3.9.1",
    "tslib": "^1.10.0"
  },
  "devDependencies": {
    "@loopback/build": "^5.4.1",
    "@loopback/eslint-config": "^7.0.1",
    "@loopback/testlab": "^3.1.5",
    "@types/node": "^10.17.24",
    "eslint": "^7.0.0",
    "pg": "^8.2.1",
    "source-map-support": "^0.5.19",
    "typescript": "~3.9.3"
  }
  
}

`
![Screenshot from 2020-05-26 12-23-10](https://user-images.githubusercontent.com/53930889/82869357-be29de80-9f4b-11ea-878b-29fa9ef5652d.png)

**Comment by madaky** [comment]
Can you try localhost at host and try

**Comment by manthan-sf** [comment]
yes i tried but got same error 
![Screenshot from 2020-05-26 16-31-47](https://user-images.githubusercontent.com/53930889/82893315-71a3ca80-9f6e-11ea-9559-6010527f6914.png)

**Comment by madaky** [comment]
Please check 

1.  npm run clean 

1. npm run build

1. npm run migrates
helps you ?
also shot a try to rm-r node_modules than npm install

**Comment by manthan-sf** [comment]
`npm run clean` , `npm run build` is working well , and making dist , but `npm run migrate` gives same error, tried 3rd point also , but problem persists

**Comment by danielalejandrin** [comment]
Hi, I got similar issue with postgresql and loopback4. I created the datasource with `lb4 datasource` command, then repository, controller and when I want to test it, I got a time out error. I've already done `npm run clean - npm run build`, deleted the node_modules folder and run npm install but I got same error. Could some one tell me what am I doing wrong?
By the way I'm following the todo tutorial of loopback page with some modifications.

![postgres_timeout](https://user-images.githubusercontent.com/58752400/82971826-41bdec80-9f99-11ea-9108-8b2063fb794b.png)
package.json
 {
  "name": "identity_provider",
  "version": "1.0.0",
  "description": "Backend for aunthentication and authorization",
  "keywords": [
    "loopback-application",
    "loopback"
  ],
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "build": "lb-tsc",
    "build:watch": "lb-tsc --watch",
    "lint": "npm run prettier:check && npm run eslint",
    "lint:fix": "npm run eslint:fix && npm run prettier:fix",
    "prettier:cli": "lb-prettier \"**/*.ts\" \"**/*.js\"",
    "prettier:check": "npm run prettier:cli -- -l",
    "prettier:fix": "npm run prettier:cli -- --write",
    "eslint": "lb-eslint --report-unused-disable-directives .",
    "eslint:fix": "npm run eslint -- --fix",
    "pretest": "npm run clean && npm run build",
    "test": "lb-mocha --allow-console-logs \"dist/__tests__\"",
    "posttest": "npm run lint",
    "test:dev": "lb-mocha --allow-console-logs dist/__tests__/**/*.js && npm run posttest",
    "docker:build": "docker build -t identity_provider .",
    "docker:run": "docker run -p 3000:3000 -d identity_provider",
    "migrate": "node ./dist/migrate",
    "prestart": "npm run build",
    "start": "node -r source-map-support/register .",
    "clean": "lb-clean dist *.tsbuildinfo .eslintcache"
  },
  "repository": {
    "type": "git"
  },
  "author": "",
  "license": "",
  "files": [
    "README.md",
    "dist",
    "src",
    "!*/__tests__"
  ],
  "dependencies": {
    "@loopback/boot": "^2.3.0",
    "@loopback/context": "^3.8.0",
    "@loopback/core": "^2.6.0",
    "@loopback/openapi-v3": "^3.4.0",
    "@loopback/repository": "^2.5.0",
    "@loopback/rest": "^5.0.0",
    "@loopback/rest-explorer": "^2.2.1",
    "@loopback/service-proxy": "^2.2.1",
    "loopback-connector-postgresql": "^3.9.1",
    "tslib": "^1.10.0"
  },
  "devDependencies": {
    "@loopback/build": "^5.4.0",
    "source-map-support": "^0.5.19",
    "@loopback/testlab": "^3.1.4",
    "@types/node": "^10.17.21",
    "@loopback/eslint-config": "^7.0.0",
    "eslint": "^7.0.0",
    "typescript": "~3.9.2"
  }
}

**Comment by danielalejandrin** [comment]
> Hi, I got similar issue with postgresql and loopback4. I created the datasource with `lb4 datasource` command, then repository, controller and when I want to test it, I got a time out error. I've already done `npm run clean - npm run build`, deleted the node_modules folder and run npm install but I got same error. Could some one tell me what am I doing wrong?
> By the way I'm following the todo tutorial of loopback page with some modifications.
> 
> ![postgres_timeout](https://user-images.githubusercontent.com/58752400/82971826-41bdec80-9f99-11ea-9108-8b2063fb794b.png)
> package.json
> {
> "name": "identity_provider",
> "version": "1.0.0",
> "description": "Backend for aunthentication and authorization",
> "keywords": [
> "loopback-application",
> "loopback"
> ],
> "main": "dist/index.js",
> "types": "dist/index.d.ts",
> "engines": {
> "node": ">=10"
> },
> "scripts": {
> "build": "lb-tsc",
> "build:watch": "lb-tsc --watch",
> "lint": "npm run prettier:check && npm run eslint",
> "lint:fix": "npm run eslint:fix && npm run prettier:fix",
> "prettier:cli": "lb-prettier "**/*.ts" "**/_.js"", "prettier:check": "npm run prettier:cli -- -l", "prettier:fix": "npm run prettier:cli -- --write", "eslint": "lb-eslint --report-unused-disable-directives .", "eslint:fix": "npm run eslint -- --fix", "pretest": "npm run clean && npm run build", "test": "lb-mocha --allow-console-logs "dist/**tests**"", "posttest": "npm run lint", "test:dev": "lb-mocha --allow-console-logs dist/**tests**/**/_.js && npm run posttest",
> "docker:build": "docker build -t identity_provider .",
> "docker:run": "docker run -p 3000:3000 -d identity_provider",
> "migrate": "node ./dist/migrate",
> "prestart": "npm run build",
> "start": "node -r source-map-support/register .",
> "clean": "lb-clean dist _.tsbuildinfo .eslintcache" }, "repository": { "type": "git" }, "author": "", "license": "", "files": [ "README.md", "dist", "src", "!_/**tests**"
> ],
> "dependencies": {
> "@loopback/boot": "^2.3.0",
> "@loopback/context": "^3.8.0",
> "@loopback/core": "^2.6.0",
> "@loopback/openapi-v3": "^3.4.0",
> "@loopback/repository": "^2.5.0",
> "@loopback/rest": "^5.0.0",
> "@loopback/rest-explorer": "^2.2.1",
> "@loopback/service-proxy": "^2.2.1",
> "loopback-connector-postgresql": "^3.9.1",
> "tslib": "^1.10.0"
> },
> "devDependencies": {
> "@loopback/build": "^5.4.0",
> "source-map-support": "^0.5.19",
> "@loopback/testlab": "^3.1.4",
> "@types/node": "^10.17.21",
> "@loopback/eslint-config": "^7.0.0",
> "eslint": "^7.0.0",
> "typescript": "~3.9.2"
> }
> }

I forgot to mention that I was using node current version 14.3.0. A suggestion of a colleague to change to LTS version 12.17.0, and it worked, no more timeout so I think it's a bug in the _loopback-connector-postgresql_ with the current version of nodejs

**Reply by sahil78623** [reply]
You can work with latest versions of node, you just need to install the latest version of "POSTGRES" connector.
The following command will do the work "npm i loopback-connector-postgresql"

**Reply by awongCM** [reply]
I'm using node version 16.0.0, still running into the same problem. And I'm using the latest Postgres connector: `^7.0.1`. 

I tried downgrading it to 12.16.3 but I'm running into npm build and npm migrate cli errors due to npm versioning compatibility issues. So it's still a chicken-and-egg situation for me right now.

Can anyone shed light on the permanent resolution for it despite using the latest Postgres connector library?

**Reply by awongCM** [reply]
Don't worry about it.

I've sorted out here - https://github.com/loopbackio/loopback-connector-postgresql/issues/636#issuecomment-1872670434

**Comment by raymondfeng** [comment]
@danielalejandrin Thank you for letting us know. Can you open a bug against `pg` - https://github.com/brianc/node-postgres/issues?

**Comment by brainafesoh** [comment]
Hi guys, hello @danielalejandrin  I'm still facing the same issue. Currently using node version **12.17.0** and the migrate command still fails.

**Comment by achrinza** [comment]
Ensure that you're using a supported version of Node.js: v10, v12, v14.

**Comment by dcalme** [comment]
Hi, 
I got the same issue, with Node LTS (14.15.1) i use n to downgrade at 12.16.2 and it works fine.

**Comment by danielalejandrin** [comment]
> @danielalejandrin Thank you for letting us know. Can you open a bug against `pg` - https://github.com/brianc/node-postgres/issues?

Sorry not to open an issue but and new with node js and I was trying a framework to use instead of express, and I drop it cause I using express with sequalize ORM

**Comment by mericano1** [comment]
I tried downgrading and it is still not working for me. Any further updates on this issue at all ?

**Reply by marioestradarosa** [reply]
@mericano1 can u provide your environment:
1. Your datasource config (_sample not showing real password etc_.)
2. your package.json (_showing only dependencies_)
3. Your nodeJS version

I will try to reproduce it today. Also note that I usually don't specify anything in the url parameter, so it remains ```url: '',```.  URL and other parameters for establishing a connection are **Mutually exclusive.**

**Reply by mericano1** [reply]
Hi Mario, thanks for your answer. I changed the code slightly and don't have this problem anymore. The code way I was doing this before (and not working) was 

```
//Application.ts

export class Application extends BootMixin(
  ServiceMixin(RepositoryMixin(RestApplication)),
) {
  constructor(options: ApplicationConfig = {}) {
    super(options);
    if (options.dsConfig) {
      const ds = this.configureDbFromOptions(options)
      this.datasource(ds) // <- This was not working
   } else {
      throw new Error("datasource must be configured")
   }

    // Set up default home page
    this.static('/', path.join(__dirname, '../public'));

    // Customize @loopback/rest-explorer configuration here
    this.configure(RestExplorerBindings.COMPONENT).to({
      path: '/explorer',
    });
    this.component(RestExplorerComponent);

    this.projectRoot = __dirname;
    // Customize @loopback/boot Booter Conventions here
    this.bootOptions = {
      controllers: {
        // Customize ControllerBooter Conventions here
        dirs: ['controllers'],
        extensions: ['.controller.js'],
        nested: true,
      },
    };
  }
}
private configureDbFromOptions(dsConfig: { url: string }) {
   console.log("configuring database from URL ")
   this.bind('datasources.config.postgres').to({
     name: 'db',
     connector: 'postgresql',
     url: dsConfig.url,
     useNewUrlParser: true,
   });
   this.bind('datasources.db').toClass(PostgresDataSource);
 }
```

The `PostgresDataSource` class is very basic


```
import {inject, lifeCycleObserver, LifeCycleObserver} from '@loopback/core';
import {juggler} from '@loopback/repository';

@lifeCycleObserver('datasource')
export class PostgresDataSource extends juggler.DataSource
  implements LifeCycleObserver {
  static dataSourceName = 'postgres';

  constructor(
    @inject('datasources.config.postgres', {optional: true})
    dsConfig: object,
  ) {
    super(dsConfig);
  }
}
```

and dependencies are

```
"dependencies": {
    "@loopback/boot": "^3.4.1",
    "@loopback/cli": "^2.21.1",
    "@loopback/core": "^2.16.1",
    "@loopback/repository": "^3.7.0",
    "@loopback/rest": "^9.3.1",
    "@loopback/rest-explorer": "^3.3.1",
    "@loopback/service-proxy": "^3.2.1",
    "@types/lodash": "^4.14.170",
    "global": "^4.4.0",
    "lodash": "^4.17.21",
    "loopback-connector-postgresql": "^5.0.1",
    "loopback4-migration": "^1.2.1",
    "moment-timezone": "^0.5.33",
    "postgres": "^1.0.2",
    "tslib": "^2.0.0"
  }
```


The main issue was that the datasource was not initialised properly. I was There is something in the lifecycle that I don't understand that well, I guess. I changed my code so that I can use different datasources from `index.ts` rather than `application.ts`.

The new main looks like this 

```
export async function main(options: ApplicationConfig = {}) {

  const ds = await getDatasourceFromEnv();
  const optionsPlusDatasource = Object.assign(options, {datasource: ds})
  const app = new Application(optionsPlusDatasource);
  app.bind('datasources.db').to(ds)
  await app.boot();
  await app.start();

  const url = app.restServer.url;
  console.log(`Server is running at ${url}`);
  console.log(`Try ${url}/ping`);

  return app;
}
```


and `getDatasourceFromEnv` is 

```
export const getDatasourceFromEnv = async () => {
  const dbHost = process.env.POSTGRES_HOST ?? 'localhost';
  const dbPort = process.env.POSTGRES_PORT ?? 5432;
  const dbUser = process.env.POSTGRES_USERNAME ?? 'postgres';
  const dbPass = process.env.POSTGRES_PASSWORD ?? 'changeme';
  const database = process.env.POSTGRES_DB ?? 'postgres';

  console.log(`Configuring database from env variables.`)
  console.log(`Connecting to host: ${dbHost} on port: ${dbPort} and database '${database}' with user: ${dbUser}`)
  const dsConfig = {
    connector: 'postgresql',
    host: dbHost,
    port: dbPort,
    user: dbUser,
    password: dbPass,
    database: database
  }
  const ds = new juggler.DataSource(dsConfig);
  return ds;
}
```

This seems to work..

**Comment by marioestradarosa** [comment]
@mericano1 please note that it is not necessary to manually code the bootstrap of _DataSources_. The DataSource Booter automatically inspects the files in `./datasources` , loads them, bind them for you.

It is better to run the command `lb4 datasource` and simple answer the questions. Then in the `./datasources` directory locate your file and change slightly the code to load config from env variables and one more important thing is the `stop` function to call `disconnect`, check [HOW the Team did it for example-shooping application](https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/datasources/redis.datasource.ts) and if you see this repository [application.ts](https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/application.ts) you will notice that no mention about datsources or datasource configuration there.

**Reply by awongCM** [reply]
https://github.com/loopbackio/loopback-next/discussions/7182#discussioncomment-7981617

I posted earlier in this chat.  Started from scratch again, tried following this guide - https://loopback.io/doc/en/lb4/Connecting-to-PostgreSQL.html, 

I'm still running into the Timeout errors again...

My local dev settings are as follows:
1. NodeJS -  16
2. Package.json
```
"dependencies": {
    "@loopback/boot": "^6.1.4",
    "@loopback/core": "^5.1.4",
    "@loopback/repository": "^6.1.4",
    "@loopback/rest": "^13.1.4",
    "@loopback/rest-explorer": "^6.1.4",
    "@loopback/service-proxy": "^6.1.4",
    "loopback-connector-postgresql": "^5.5.2",
    "tslib": "^2.0.0"
  },
  "devDependencies": {
    "@loopback/build": "^10.1.4",
    "@loopback/eslint-config": "^14.0.5",
    "@loopback/testlab": "^6.1.4",
    "@types/node": "^16.18.68",
    "eslint": "^8.51.0",
    "source-map-support": "^0.5.21",
    "typescript": "~5.2.2"
  }
```

3. DataSource config

```
const config = {
  name: 'ExpenseTrackerBeDataSource',
  connector: 'postgresql',
  url: '',
  host: 'localhost',
  port: 5432,
  user: '',
  password: '',
  database: 'postgres',
  debug: true,
  ssl: false
};

```

I confirmed the above postgres db connectivity settings works when using DB client like TablePlus.  I can connect to my db without username/password supplied.

**Reply by awongCM** [reply]
@marioestradarosa 
https://github.com/loopbackio/loopback-next/discussions/7182#discussioncomment-968164
When you get a chance, can you please to assist the issue I'm experiencing?  I want to continue getting on Postgres connector plugin to work  so I can start working on exploring the integrating db engine on my local machine with loopback.

Thank you, much appreciated.

**Comment by awongCM** [comment]
https://github.com/loopbackio/loopback-next/discussions/7182#discussioncomment-7978238

As per the question post above, has anyone else come up with a resolution in getting loopback to work with Postgres properly?

**Reply by awongCM** [reply]
As explained here - https://github.com/loopbackio/loopback-connector-postgresql/issues/636#issuecomment-1872670434

Please ignore this.
